Add description and title to consumables

1drturtle opened the pull request:
Summary
======
Adds a description and title to custom counters/consumables
In the title, `[name]` is replaced with the character name and `[cname]` is replace with the custom counter name.

Resolves #1276 (AFR-639)

Methods added
----
`AliasCustomCounter` -
* `AliasCustomCounter.title`
* `AliasCustomCounter.desc`

`CustomCounter` -
* `CustomCounter.title`
* `CustomCounter.desc`


Checklist
======
<!-- Put an "x" inside the braces to tick checkboxes, e.g. [x]. -->
#### PR Type
<!-- If the PR closes an issue, mention the issue at the top of the PR with "Resolves #X". -->
- [X] This PR is a code change that implements a feature request.
- [ ] This PR fixes an issue.
- [ ] This PR adds a new feature that is not an open feature request.
- [ ] This PR is not a code change (e.g. documentation, README, ...)
#### Other
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [X] If code changes were made then they have been tested.
- [X] I have updated the documentation to reflect the changes.

mommothazaz123 commented on cogs5e/models/sheet/player.py:
```diff
@@ -94,9 +94,13 @@ class CustomCounter:
 
     def __init__(self, character, name, value, minv=None, maxv=None, reset=None, display_type=None, live_id=None,
-                 reset_to=None, reset_by=None):
+                 title=None, desc=None, reset_to=None, reset_by=None):
```
```suggestion
                 reset_to=None, reset_by=None, title=None, desc=None):
```

I don't think `__init__` is ever called with only positional args, but this prevents breaking it if it is.

mommothazaz123 commented on cogs5e/models/sheet/player.py:
```diff
@@ -116,12 +120,12 @@ def from_dict(cls, char, d):
     @classmethod
     def new(cls, character, name, minv=None, maxv=None, reset=None, display_type=None, live_id=None,
-            reset_to=None, reset_by=None):
+            title=None, desc=None, reset_to=None, reset_by=None):
```
```suggestion
            reset_to=None, reset_by=None, title=None, desc=None):
```

mommothazaz123 commented on cogs5e/models/sheet/player.py:
```diff
@@ -168,7 +172,18 @@ def new(cls, character, name, minv=None, maxv=None, reset=None, display_type=Non
+            raise InvalidArgument('Name must be less than 256 characters.')
+
+        return cls(character, name.strip(), initial_value, minv, maxv, reset, display_type, live_id,
+                   title, desc, reset_to, reset_by)
```
```suggestion
                   reset_to, reset_by, title, desc)
```
to match the positional arg suggestions

mommothazaz123 commented on cogs5e/models/sheet/player.py:
```diff
@@ -196,6 +211,25 @@ def get_reset_to(self):
     def value(self):
         return self._value
 
+    @property
```
I'm not entirely sure these properties (and similarly, `_title` and `_desc` being private attrs) are entirely necessary.

As is, `CustomCounter.title` (the property here) is what is called when it's committed to the database, so the `[cname]` and `[name]`s are lost - I don't believe `[cname]` and `[name]` are necessary, though.

mommothazaz123 commented on cogs5e/models/sheet/player.py:
```diff
@@ -274,14 +312,15 @@ def full_str(self):
     def __str__(self):
         _max = self.get_max()
 
+        out = self.desc + '\n' if self.desc else ''
```
This'll put the description (up to 800 characters) in *every* location the counter is stringified. That doesn't seem necessary - for example, if I had an alias 
```
!alias wildmagic embed <drac2>
out = [...]
# some wild magic code here
out.append(f'-f "Some Resource|{{character().cc_str('Some Resource')}}"')
return '\n'.join(out)
```

I wouldn't want the description to appear.

mommothazaz123 reviewed (changes requested):
Thanks for the contribution! I'm not entirely sure what the scope of the feature request this PR is closing is. Should titles and descriptions appear everywhere? In rests? Only when they're manually changed with `!cc`?

I believe the original scope of the FR only encompasses the latter (and thus that the title and desc handling should be in that command rather than in methods on the cc object), and that the `[name]` and `[cname]` modification is unnecessary.

1drturtle commented:
Updated - 
Removed description from `__str__`
Fixed `title` and `desc` properties (`[name]` and `[cname]` are still supported, but are replaced in the command instead.)
Moved order of `title` and `desc`
![image](https://user-images.githubusercontent.com/16567386/97043320-4d114600-1540-11eb-9fc0-b1eadf774e26.png)
(In aliasing)
![image](https://user-images.githubusercontent.com/16567386/97043447-7df17b00-1540-11eb-99da-58a18befbfa9.png)

RFC - Remove `[cname]` since when it's created they know the name of it?

mommothazaz123 commented:
<img width="600" alt="Screenshot 2020-10-26 at 11 43 51" src="https://user-images.githubusercontent.com/8866981/97214589-8d5f0700-1780-11eb-8fb0-08d3a88053b0.png">

Maybe the description should have its own field (under the meta of cc info/cc value) - this looks a little cluttered

1drturtle commented:
With or without description header? I'm leaning more towards with header so I that's what I committed.
![image](https://user-images.githubusercontent.com/16567386/97233111-c74a0080-17b4-11eb-93c5-0ceb43ecaae3.png)

mommothazaz123 commented on cogs5e/models/sheet/player.py:
```diff
@@ -252,11 +267,13 @@ def full_str(self):
 
+        val = ''
+
         if self.display_type == 'bubble':
             assert self.max is not None
-            val = f"{bubble_format(self.value, _max)}\n"
+            val += f"{bubble_format(self.value, _max)}\n"
         else:
-            val = f"**Current Value**: {self.value}\n"
+            val += f"**Current Value**: {self.value}\n"
```
This change can probably be reverted

mommothazaz123 commented on cogs5e/gametrack.py:
```diff
@@ -434,15 +434,25 @@ async def spellbook_remove(self, ctx, *, spell_name):
+        # replace [name] in title
+        cc_embed_title = cc_embed_title.replace('[name]', character.name)
+
+        cc_desc = counter.desc if counter.desc else ''
```
The use of `cc_desc` below can probably be changed to just checks on `counter.desc`, and this variable removed
